main/src/ca/uwaterloo/flix/language/phase/Instances.scala

```diff
@@ -27,19 +28,154 @@ import ca.uwaterloo.flix.util.{InternalCompilerException, ParOps, Validation}
 object Instances extends Phase[TypedAst.Root, TypedAst.Root] {
 
   /**
-    * Validates instances in the given AST root.
+    * Validates instances and classes in the given AST root.
     */
   override def run(root: TypedAst.Root)(implicit flix: Flix): Validation[TypedAst.Root, CompilationError] = flix.phase("Instances") {
     for {
       _ <- visitInstances(root)
+      _ <- visitClasses(root)
     } yield root
   }
 
+  /**
+    * Validates all instances in the given AST root.
+    */
+  private def visitClasses(root: TypedAst.Root)(implicit flix: Flix): Validation[Unit, InstanceError] = {
+
+    /**
+      * Checks that every super class of `class0` is lawful, unless `class0` is marked `lawless`.
+      */
+    def checkLawfulSuperClasses(class0: TypedAst.Class): Validation[Unit, InstanceError] = class0 match {
+      case TypedAst.Class(_, mod, sym, _, superClasses, _, _, loc) =>
+        if (mod.isLawless) {
+          ().toSuccess
+        } else {
+          Validation.traverseX(superClasses) {
+            superSym =>
+              val superClass = root.classes(superSym)
+              if (superClass.mod.isLawless) {
+                InstanceError.LawlessSuperClass(sym, superSym, loc).toFailure
+              } else {
+                ().toSuccess
+              }
+          }
+        }
+    }
+
+    /**
+      * Creates a list of all the sigs used in the given `defn0`.
+      */
+    def findUsedSigs(defn0: TypedAst.Def): List[Symbol.SigSym] = {
```

Comment:
I'm going to revisit this to subtract from a target list of sigs, rather than to build a list of used sigs.

Reply:
Or perhaps instead we should have a `TypedAstOps` where we can share functionality like this?

Reply:
Also it might be both clearer and faster to return a set of SigSym.

Reply:
If you could change `findUsedSigs` to return a set and move it into `TypedAstOps` and possibly rename it to `sigSymsOf` then we I think we are good to go!

Reply:
I'm wondering if this could be generalized into some kind of fold, since I imagine finding sigSyms specifically would be rare, while collecting something in general might be useful elsewhere.

Reply:
We could implement folds on all the Asts, but when doing the thought experiment I came to the conclusion that often a fold is not enough.